R/tmb-methods.R

```diff
@@ -427,3 +427,50 @@ h_residuals_response <- function(object) {
   assert_class(object, "mmrm_tmb")
   component(object, "y_vector") - unname(fitted(object))
 }
+
+
+#' @describeIn mmrm_tmb_methods Simulates from an mmrm model object
+#' @importFrom stats simulate
+#' @exportS3Method
+#'
+#' @param object an object of class 'mmrm'
+#' @param nsim number of response vectors to simulate. Defaults to 1.
+#' @param seed an object specifying if and how the random number generator should be initialized (‘seeded’).
+#' @param newdata data frame for which to evaluate predictions.
+#' @param ...
+#'
+#' @return A \code{data.frame} of dimension [n, m] where n is the number of observations
+#' fitted in the model, and m is the number \code{nsim} of simulated responses.
+simulate.mmrm_tmb <- function(object, nsim = 1,
+                          seed = NULL, newdata = object$data,
+                          ...){
+  if(!is.null(seed)) set.seed(seed)
+
+  # build data.frame of prediction vectors
+  ret <- predict(object, newdata = newdata, se.fit = FALSE, interval = "none")
+  ret <- data.frame(replicate(nsim, ret))
+  covmat <- VarCorr(object)
+  grouped <- if(object$tmb_data$n_groups > 1) TRUE else FALSE
+
+  for(i in 1:object$tmb_data$n_subjects){
+
+    # Obtain indices of data.frame belonging to subject i
+    num_visits <- object$tmb_data$subject_n_visits[i]
+    inds <- (object$tmb_data$subject_zero_inds[i] + 1) : (object$tmb_data$subject_zero_inds[i] + num_visits)
+    visits <- object$tmb_data$visits_zero_inds[inds] + 1
+
+    # get relevant covariance matrix for subject i
+    covmat_i <- if(grouped){
+      covmat[[object$tmb_data$subject_groups[i]]]
+    }else{
+      covmat
+    }
+    # subset covariance matrix
+    covmat_i <- covmat_i[visits, visits]
```

Comment:
this covariance is not the conditional covariance matrix given some observed response

Reply:
Do we have the conditional covariance matrices calculated inside `predict()` somewhere? Could we expose them in the returned list (maybe optional) or use internal `predict` helpers to get them here?

Reply:
yes it is calculated at https://github.com/openpharma/mmrm/blob/d485c17240fe960a163fca9e6928925a8c7b9c1b/src/predict.cpp#L102

please note that only the diagonal is used here.

Since the data is not structured (not ordered probably) so maybe it is good to return the full matrix (still might be too large a matrix).  (sparse matrix might be a good option?)

Reply:
Ah ok that is very interesting. That refers back to what I asked above re: are we using everything we can from `predict()` - so it seems we could return a bit more from C++ side to make the calculation easy here?
And re: storage- I see that we have this for each individual, I guess then we can just return the full matrix for each individual. Sparse matrix won't help I think because we don't have structure within individuals (at least not with unstructured cov matrix which is typically used)

Reply:
I mean, with sparse matrix, even if we return the full matrix we don't pay much; the elements between individuals are still missing; the storage is not that much (for each element we store the row/column index). With that we can make it clear if the input data is not correctly ordered.

If we provide the full matrix for each individual, while this works and could be even simpler, we need more documentation there that the returned list of matrix does not have the same order of original input data  (maybe also return the index for each subjects too)